demo/Render.hpp

```diff
@@ -0,0 +1,841 @@
+
+            float ambientStrength = 0.75;
+            vec3 ambient = ambientStrength * lightColor;
+
+            float diff = max(dot(fragNormal, light), 0.0);
+            vec3 diffuse = diff * lightColor;
+
+            float specularStrength = 0.5;
+            vec3 viewDir = normalize(eye - fragPos);
+            vec3 reflectDir = reflect(-light, fragNormal);
+            float spec = pow(max(dot(viewDir, reflectDir), 0.0), 16);
+            vec3 specular = specularStrength * spec * lightColor;
+
+            fragColor = vec4((ambient + diffuse + specular) * color, 1.0);
+        }
+    )";
+
+    /**
+     * @brief Construct default intialized render instance
+     * @attention must be called from the main thread
+     */
+    Renderer();
+
+    ~Renderer();
+
+    /**
+     * @brief Initializes glfw library
+     * @attention must be called from the main thread
+     */
+    void InitializeGlfw();
+
+    /**
+     * @brief Initializes glfw context
+     * @attention must be called from the main thread
+     */
+    void InitializeContext();
+
+    /**
+     * @brief Initializes glfw callbacks
+     * @attention must be called from the main thread
+     */
+    void InitializeCallbacks() const;
+
+    /**
+     * @brief Compiles and links the shader program
+     */
+    void InitializeShaderProgram();
+
+    /**
+     * @brief Resize callback function
+     * @param[in] window poiter to the current glfw context window
+     * @param[in] width new width of the window
+     * @param[in] height
+     */
+    static void Resize(GLFWwindow* window, int width, int height);
+
+    /**
+     * @brief KeyButton callback function
+     * @param[in] window poiter to the current glfw context window
+     * @param[in] key button id
+     * @param[in] scancode
+     * @param[in] action action type
+     * @param[in] mods
+     */
+    static void KeyButton(GLFWwindow* window, int key, int scancode, int action, int mods);
+
+    /**
+     * @brief CursorMove callback function
+     * @param[in] window poiter to the current glfw context window
+     * @param[in] xpos new mouse x value in the window coordinates
+     * @param[in] ypos new mouse y value in the window coordinates
+     */
+    static void CursorMove(GLFWwindow* window, double xpos, double ypos);
+};
+
+namespace primitive
+{
+/**
+ * @brief Hight level primitive initialization class for the render
+ */
+class Primitive
+{
+public:
+    /**
+     * @brief Allocates space for the mesh data in the renderer
+     */
+    Primitive();
+
+    ~Primitive();
+
+    /**
+     * @brief Sets model matrix for the given primitive's mesh
+     * @param model stores local to world space mesh transformation stores local to world space mesh transformation
+     */
+    void SetModel(glm::mat4 model) const;
+
+    /**
+     * @brief Returns model matrix of the mesh
+     * @return model matrix
+     */
+    glm::mat4 GetModel() const;
+
+protected:
+    bool m_initialized;
```

Comment:
I didn't quite get it, what does this variable do? As far as I see, it's always `true`.